File: frontend/src/api/dienstbuchPlaner.ts
import { apiDelete, apiGet, apiPatch, apiPost, apiUpload } from "./client";
import type {
  DiveraUebertragungErgebnis,
  FeiertagOut,
  PlanerKategorieOut,
  PlanTerminEreignisOut,
  PlanTerminOut,
  PlanVorlageOut,
  VorlageUeberfaelligOut,
} from "./types";

// --- Kategorien ----------------------------------------------------------

export interface PlanerKategorieAnlegen {
  name: string;
  farbe: string;
  reihenfolge?: number;
  aktiv?: boolean;
}

export const holeKategorien = () =>
  apiGet<PlanerKategorieOut[]>("/dienstbuch-planer/kategorien");

export const legeKategorieAn = (daten: PlanerKategorieAnlegen) =>
  apiPost<PlanerKategorieOut>("/dienstbuch-planer/kategorien", daten);

export const aktualisiereKategorie = (id: number, daten: Partial<PlanerKategorieAnlegen>) =>
  apiPatch<PlanerKategorieOut>(`/dienstbuch-planer/kategorien/${id}`, daten);

// --- Vorlagen --------------------------------------------------------------

export interface PlanVorlageAnlegen {
  titel: string;
  beschreibung?: string | null;
  wiederholungstyp: string;
  intervall?: number | null;
  wochentag?: number | null;
  kalenderwoche?: number | null;
  kw_paritaet?: "gerade" | "ungerade" | null;
  mindest_intervall_aktiv?: boolean;
  mindest_intervall_tage?: number | null;
  startdatum: string;
  enddatum?: string | null;
  uhrzeit?: string | null;
  endzeit?: string | null;
  aktiv?: boolean;
  kategorie_ids?: number[];
}

export const holeVorlagen = () => apiGet<PlanVorlageOut[]>("/dienstbuch-planer/vorlagen");

export const legeVorlageAn = (daten: PlanVorlageAnlegen) =>
  apiPost<PlanVorlageOut>("/dienstbuch-planer/vorlagen", daten);

export const aktualisiereVorlage = (id: number, daten: Partial<PlanVorlageAnlegen>) =>
  apiPatch<PlanVorlageOut>(`/dienstbuch-planer/vorlagen/${id}`, daten);

export const loescheVorlage = (id: number) =>
  apiDelete<void>(`/dienstbuch-planer/vorlagen/${id}`);

// --- Termine ----------------------------------------------------------------

export const holeTermine = (jahr: number) =>
  apiGet<PlanTerminOut[]>("/dienstbuch-planer/termine", { jahr });

export const stelleJahrSicher = (jahr: number) =>
  apiPost<PlanTerminOut[]>(`/dienstbuch-planer/termine/jahr/${jahr}/sicherstellen`);

export interface PlanPlatzhalterAnlegen {
  titel: string;
  beschreibung?: string | null;
  jahr: number;
  kategorie_ids?: number[];
}

export const legePlatzhalterAn = (daten: PlanPlatzhalterAnlegen) =>
  apiPost<PlanTerminOut>("/dienstbuch-planer/platzhalter", daten);

export interface PlanTerminAnlegen {
  titel: string;
  beschreibung?: string | null;
  zieldatum: string;
  uhrzeit?: string | null;
  endzeit?: string | null;
  kategorie_ids?: number[];
}

export const legeTerminAn = (daten: PlanTerminAnlegen) =>
  apiPost<PlanTerminOut>("/dienstbuch-planer/termine", daten);

export interface PlanTerminAktualisieren {
  titel?: string;
  beschreibung?: string | null;
  zieldatum?: string | null;
  uhrzeit?: string | null;
  endzeit?: string | null;
  kategorie_ids?: number[];
}

export const aktualisiereTermin = (id: number, daten: PlanTerminAktualisieren) =>
  apiPatch<PlanTerminOut>(`/dienstbuch-planer/termine/${id}`, daten);

export const bestaetigeTermin = (id: number) =>
  apiPost<PlanTerminOut>(`/dienstbuch-planer/termine/${id}/bestaetigen`);

export const setzeTerminAufEntwurf = (id: number) =>
  apiPost<PlanTerminOut>(`/dienstbuch-planer/termine/${id}/entwurf`);

export const holeTerminEreignisse = (id: number) =>
  apiGet<PlanTerminEreignisOut[]>(`/dienstbuch-planer/termine/${id}/ereignisse`);

export const holeUeberfaelligeVorlagen = () =>
  apiGet<VorlageUeberfaelligOut[]>("/dienstbuch-planer/ueberfaellig");

// --- Feiertage (Phase 2) ---------------------------------------------------

export const holeFeiertage = (jahr: number) =>
  apiGet<FeiertagOut[]>("/dienstbuch-planer/feiertage", { jahr });

export const holeBundeslaender = () =>
  apiGet<Record<string, string>>("/dienstbuch-planer/feiertage/bundeslaender");

export const legeFeiertagAn = (datum: string, name: string) =>
  apiPost<FeiertagOut>("/dienstbuch-planer/feiertage", { datum, name });

export const loescheFeiertag = (id: number) =>
  apiDelete<void>(`/dienstbuch-planer/feiertage/${id}`);

/** Baut die gesetzlichen Feiertage eines Jahres neu auf (nach Bundesland-Wechsel). */
export const seedeFeiertage = (jahr: number) =>
  apiPost<{ eingefuegt: number }>(`/dienstbuch-planer/feiertage/seed?jahr=${jahr}`);

// --- Excel (Phase 3) -------------------------------------------------------

export async function ladeJahresExport(jahr: number): Promise<void> {
  const blob = await apiGet<Blob>(`/dienstbuch-planer/export.xlsx?jahr=${jahr}`);
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `dienstplan-${jahr}.xlsx`;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}

export interface ImportErgebnis {
  angelegt: number;
  uebersprungen: number;
  fehler: { zeile: number; fehler: string }[];
}

export const importiereJahr = (jahr: number, datei: File) =>
  apiUpload<ImportErgebnis>(`/dienstbuch-planer/import?jahr=${jahr}`, datei, "datei");

// --- Divera (Phase 4) ------------------------------------------------------

export interface DiveraGruppe {
  id: number;
  name: string;
}

export interface DiveraInfo {
  aktiv: boolean;
  gruppen: DiveraGruppe[];
}

export const holeDiveraInfo = () => apiGet<DiveraInfo>("/dienstbuch-planer/divera-info");

export interface DiveraUebertragung {
  termin_ids: number[];
  gruppen_ids?: number[];
  erinnerung_minuten?: number | null;
  send_push?: boolean;
}

export const uebertrageAnDivera = (daten: DiveraUebertragung) =>
  apiPost<DiveraUebertragungErgebnis[]>("/dienstbuch-planer/divera-uebertragen", daten);

File: backend/app/schemas/dienstbuch_planer.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

HEX_FARBE_PATTERN = r"^#[0-9A-Fa-f]{6}$"

WIEDERHOLUNGSTYPEN = (
    "jaehrlich",
    "monatlich",
    "alle_x_tage",
    "alle_x_wochen",
    "alle_x_monate",
    "alle_x_jahre",
)


class PlanerKategorieAnlegen(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    farbe: str = Field(pattern=HEX_FARBE_PATTERN)
    reihenfolge: int = 0
    aktiv: bool = True


class PlanerKategorieAktualisieren(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    farbe: str | None = Field(default=None, pattern=HEX_FARBE_PATTERN)
    reihenfolge: int | None = None
    aktiv: bool | None = None


class PlanerKategorieOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    farbe: str
    reihenfolge: int
    aktiv: bool


class PlanVorlageAnlegen(BaseModel):
    titel: str = Field(min_length=1, max_length=255)
    beschreibung: str | None = None
    wiederholungstyp: str = Field(pattern="|".join(WIEDERHOLUNGSTYPEN))
    intervall: int | None = Field(default=None, ge=1)
    wochentag: int | None = Field(default=None, ge=0, le=6)
    kalenderwoche: int | None = Field(default=None, ge=1, le=53)
    kw_paritaet: str | None = Field(default=None, pattern="gerade|ungerade")
    mindest_intervall_aktiv: bool = False
    mindest_intervall_tage: int | None = Field(default=None, ge=1)
    startdatum: date
    enddatum: date | None = None
    aktiv: bool = True
    kategorie_ids: list[int] = []


class PlanVorlageAktualisieren(BaseModel):
    titel: str | None = Field(default=None, min_length=1, max_length=255)
    beschreibung: str | None = None
    wiederholungstyp: str | None = Field(default=None, pattern="|".join(WIEDERHOLUNGSTYPEN))
    intervall: int | None = Field(default=None, ge=1)
    wochentag: int | None = Field(default=None, ge=0, le=6)
    kalenderwoche: int | None = Field(default=None, ge=1, le=53)
    kw_paritaet: str | None = Field(default=None, pattern="gerade|ungerade")
    mindest_intervall_aktiv: bool | None = None
    mindest_intervall_tage: int | None = Field(default=None, ge=1)
    startdatum: date | None = None
    enddatum: date | None = None
    aktiv: bool | None = None
    kategorie_ids: list[int] | None = None


class PlanVorlageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    titel: str
    beschreibung: str | None
    wiederholungstyp: str
    intervall: int | None
    wochentag: int | None
    kalenderwoche: int | None
    kw_paritaet: str | None
    mindest_intervall_aktiv: bool
    mindest_intervall_tage: int | None
    startdatum: date
    enddatum: date | None
    aktiv: bool
    kategorien: list[PlanerKategorieOut]


class PlanPlatzhalterAnlegen(BaseModel):
    titel: str = Field(min_length=1, max_length=255)
    beschreibung: str | None = None
    jahr: int = Field(ge=2000, le=2200)
    kategorie_ids: list[int] = []


class PlanTerminAktualisieren(BaseModel):
    titel: str | None = Field(default=None, min_length=1, max_length=255)
    beschreibung: str | None = None
    zieldatum: date | None = None
    ist_platzhalter: bool | None = None
    kategorie_ids: list[int] | None = None


class PlanTerminOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    vorlage_id: int | None
    vorlage_titel: str | None = None
    jahr: int
    titel: str
    beschreibung: str | None
    zieldatum: date | None
    ist_platzhalter: bool
    status: str
    dienstbuch_id: int | None
    dienstbuch_erzeugt_am: datetime | None
    kategorien: list[PlanerKategorieOut]


class PlanTerminEreignisOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    zeitpunkt: datetime
    typ: str
    beschreibung: str
    akteur_name: str | None


class VorlageUeberfaelligOut(BaseModel):
    vorlage_id: int
    titel: str
    letztes_zieldatum: date | None
    tage_ueberfaellig: int
